### ISSUE TITLE
ca-cert

### ISSUE THREAD

**Issue by ukos-git** [ioistired/ps3-update-dl#2]
After install ing the program via pip, I get some errors regarding the stored certificate.

```console
user@ubuntu:$ ~/.local/bin/ps3udl -o update BCES00795
Traceback (most recent call last):
  File "/home/user/.local/bin/ps3udl", line 8, in <module>
    sys.exit(main())
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 189, in main
    _main()
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 184, in _main
    download_updates(**kwargs, title_id=title_id)
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 131, in download_updates
    (title, *_), updates = download_info(title_id)
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 67, in download_info
    r = session.get(URL_FORMAT.format(id=title_id))
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 416, in send
    self.cert_verify(conn, request.url, verify, cert)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 227, in cert_verify
    raise IOError("Could not find a suitable TLS CA certificate bundle, "
OSError: Could not find a suitable TLS CA certificate bundle, invalid path: /home/user/.local/lib/python3.8/site-packages/playstation-ca.crt

user@ubuntu:$ wget https://raw.githubusercontent.com/iomintz/ps3-update-dl/trunk/playstation-ca.crt -O /home/user/.local/lib/python3.8/site-packages/playstation-ca.crt
--2021-03-07 15:19:32--  https://raw.githubusercontent.com/iomintz/ps3-update-dl/trunk/playstation-ca.crt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2908 (2,8K) [text/plain]
Saving to: ‘/home/user/.local/lib/python3.8/site-packages/playstation-ca.crt’

/home/user/.local/lib/python3.8/site-packages/playstat 100%[======================================================================================================================================>]   2,84K  --.-KB/s    in 0s      

2021-03-07 15:19:32 (15,9 MB/s) - ‘/home/user/.local/lib/python3.8/site-packages/playstation-ca.crt’ saved [2908/2908]

user@ubuntu:$ ~/.local/bin/ps3udl -o update BCES00795
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 488, in wrap_socket
    cnx.do_handshake()
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1934, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1671, in _raise_ssl_error
    _raise_current_error()
  File "/usr/lib/python3/dist-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 670, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 381, in _make_request
    self._validate_conn(conn)
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 978, in _validate_conn
    conn.connect()
  File "/usr/lib/python3/dist-packages/urllib3/connection.py", line 361, in connect
    self.sock = ssl_wrap_socket(
  File "/usr/lib/python3/dist-packages/urllib3/util/ssl_.py", line 377, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 494, in wrap_socket
    raise ssl.SSLError("bad handshake: %r" % e)
ssl.SSLError: ("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')])",)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 724, in urlopen
    retries = retries.increment(
  File "/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 439, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='a0.ww.np.dl.playstation.net', port=443): Max retries exceeded with url: /tpl/np/BCES00795/BCES00795-ver.xml (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')])")))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/.local/bin/ps3udl", line 8, in <module>
    sys.exit(main())
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 189, in main
    _main()
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 184, in _main
    download_updates(**kwargs, title_id=title_id)
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 131, in download_updates
    (title, *_), updates = download_info(title_id)
  File "/home/user/.local/lib/python3.8/site-packages/ps3_update_dl.py", line 67, in download_info
    r = session.get(URL_FORMAT.format(id=title_id))
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 514, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='a0.ww.np.dl.playstation.net', port=443): Max retries exceeded with url: /tpl/np/BCES00795/BCES00795-ver.xml (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')])")))
```

The certificate that I would get via openssl is
```console
user@ubuntu:$ openssl s_client -showcerts -connect a0.ww.np.dl.playstation.net:443
CONNECTED(00000003)
depth=1 C = JP, O = Sony Computer Entertainment Inc., CN = SCEI DNAS Root 05
verify error:num=19:self signed certificate in certificate chain
verify return:1
depth=0 C = JP, ST = Tokyo, L = Minato-ku, O = Sony Interactive Entertainment Inc., OU = Network Engineering & Operation Dept., CN = *.ww.np.dl.playstation.net
verify error:num=68:CA signature digest algorithm too weak
verify return:1
depth=1 C = JP, O = Sony Computer Entertainment Inc., CN = SCEI DNAS Root 05
verify return:1
depth=0 C = JP, ST = Tokyo, L = Minato-ku, O = Sony Interactive Entertainment Inc., OU = Network Engineering & Operation Dept., CN = *.ww.np.dl.playstation.net
verify return:1
---
Certificate chain
 0 s:C = JP, ST = Tokyo, L = Minato-ku, O = Sony Interactive Entertainment Inc., OU = Network Engineering & Operation Dept., CN = *.ww.np.dl.playstation.net
   i:C = JP, O = Sony Computer Entertainment Inc., CN = SCEI DNAS Root 05
-----BEGIN CERTIFICATE-----
MIIEUjCCAzqgAwIBAgICEOIwDQYJKoZIhvcNAQEFBQAwVDELMAkGA1UEBhMCSlAx
KTAnBgNVBAoTIFNvbnkgQ29tcHV0ZXIgRW50ZXJ0YWlubWVudCBJbmMuMRowGAYD
VQQDExFTQ0VJIEROQVMgUm9vdCAwNTAeFw0yMDA4MjQwNTUwMDNaFw0yMzA4MjQw
NTUwMDNaMIG0MQswCQYDVQQGEwJKUDEOMAwGA1UECAwFVG9reW8xEjAQBgNVBAcM
CU1pbmF0by1rdTEsMCoGA1UECgwjU29ueSBJbnRlcmFjdGl2ZSBFbnRlcnRhaW5t
ZW50IEluYy4xLjAsBgNVBAsMJU5ldHdvcmsgRW5naW5lZXJpbmcgJiBPcGVyYXRp
b24gRGVwdC4xIzAhBgNVBAMMGioud3cubnAuZGwucGxheXN0YXRpb24ubmV0MIIB
IjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArNEgHiKWkByRNvGoyZ3OdvvB
GCBeIkp+3q7sMSpdwACFD961FNXRgbm+tz5J3sMolffq2nPP32sbwI9m8XEOxamn
9pN49diMUOfLhgCygRG7tCnfqOfIHGGRpTPBg9fLnENscfPbyuND56Uw6IpRYHdU
QyTf8vxyU7h36oB4ydIPr5sIxXVfCVjweuWPhqWf0uBNG9QqnRMD8pNz67qkK1Uf
tz2KwKZFRp6Ifkx33yK17jW1KihjlqzWTW+RUw7Pt3Mzp/5AOzzR5AzNBV/SsB8e
cVntoC67M6UaIqnxMTs9RSp/XNTbbGN9WVkUOtLTctv++7KJrjUSkHxct2OV+QID
AQABo4HMMIHJMAkGA1UdEwQCMAAwHwYJYIZIAYb4QgENBBIWEEROQVMgQ2VydGlm
aWNhdGUwHQYDVR0OBBYEFOG0p8ykxAQAMwD+t5z8zmVvQsoKMHwGA1UdIwR1MHOA
FMZWoTNbT86Dd2IFRIZtIFe1r9rcoVikVjBUMQswCQYDVQQGEwJKUDEpMCcGA1UE
ChMgU29ueSBDb21wdXRlciBFbnRlcnRhaW5tZW50IEluYy4xGjAYBgNVBAMTEVND
RUkgRE5BUyBSb290IDA1ggEAMA0GCSqGSIb3DQEBBQUAA4IBAQCSkXVrAyevSJe5
lfA+MFN2Iv0WrYytEGPT8LIJJIYX6irMpHJvPTAVd4rGmYf4D6PCQht0ua8xGEML
3OlQP6QKYlWGmAVKQFw5Hh81GRhSgSb7gSt9PfeCZ7UhS8xa35HlR5vOlApWvwry
leSqvEGT96QGSj9LU1CF2gXf2PjOuyIf/J4r9QEex41ta8gaoz3/WHQRae70OVBH
kdkOxqCZW/jqPS6gtbSHytUNRb8vJHvQyiWVSur5z0PaHyWfZrBE7Mqv+OyiZqyZ
7N67DLbeO/ZOsH1gJLZ+XhXgoukgsZJkwKpnLaMcIdNULSWmDVmA1iGq+3pwSFpz
xtALjS5/
-----END CERTIFICATE-----
 1 s:C = JP, O = Sony Computer Entertainment Inc., CN = SCEI DNAS Root 05
   i:C = JP, O = Sony Computer Entertainment Inc., CN = SCEI DNAS Root 05
-----BEGIN CERTIFICATE-----
MIID0jCCArqgAwIBAgIBADANBgkqhkiG9w0BAQUFADBUMQswCQYDVQQGEwJKUDEp
MCcGA1UEChMgU29ueSBDb21wdXRlciBFbnRlcnRhaW5tZW50IEluYy4xGjAYBgNV
BAMTEVNDRUkgRE5BUyBSb290IDA1MB4XDTA0MDcxMjA5MDExOVoXDTM3MTIwNjA5
MDExOVowVDELMAkGA1UEBhMCSlAxKTAnBgNVBAoTIFNvbnkgQ29tcHV0ZXIgRW50
ZXJ0YWlubWVudCBJbmMuMRowGAYDVQQDExFTQ0VJIEROQVMgUm9vdCAwNTCCASIw
DQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANmPeza8PwCqlI7esOGIkoSESnIN
g72ZD3Ut63jy7SdothPIvGBqVZWYkIpqJYJd1I4Nh//IpXQCQL0PnJLrh9BBeowq
Muf5NNq3Us80Ihiu9CvNEAEO18g3OFV1TYdSwQ5zUsk33OUeI7h4aBPDVcZXYeHt
dbPLqe4K8igian5prrAD5S6h28t8aAm+qMWRo+bW25B/841XwDGBP7/IxZv8Yoio
rCo80CVYe6lGoU08eeqQiaHI5zAF281DWZSoVfLjJUEWmEnxqr8aOhszRGePi+Ei
7UQjHDuZX9rLhDI1zAND+BA259tn/iwOqVXe20OccJllHJcG4Ecmd98f5qMCAwEA
AaOBrjCBqzAdBgNVHQ4EFgQUxlahM1tPzoN3YgVEhm0gV7Wv2twwfAYDVR0jBHUw
c4AUxlahM1tPzoN3YgVEhm0gV7Wv2tyhWKRWMFQxCzAJBgNVBAYTAkpQMSkwJwYD
VQQKEyBTb255IENvbXB1dGVyIEVudGVydGFpbm1lbnQgSW5jLjEaMBgGA1UEAxMR
U0NFSSBETkFTIFJvb3QgMDWCAQAwDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQUF
AAOCAQEACZPihjwXA27wJ03tEKcHAeFLi8aBw2ysH4GwuH1dWb3UpuznWOB0iQT1
wQocnEFYCJx5XFEnj4aLWpSHLEq/sSO+my+aPoTEsy20ajF+YLYZm0bZxH50CJYh
rkET4C2aC0XvhGp9k1JQ1o0W6+cFT5LTlXapsq8Btt31t+XDPX7RqGV4WGekt3hM
T7xRc7JWXdAQijIrbYi8mtbM07KEGnPU6IT8C47+0mSurpwLOoWL1tPgo6ePpLNi
c4quUMgh9RXVjeTyXOMmyYdeUm2gt7qErvQONli+6Epmhm0A2khpIMHSpQjTE8gV
rZp42a6+zg1iYy2vFBOmiQ17GRUl0A==
-----END CERTIFICATE-----
---
Server certificate
subject=C = JP, ST = Tokyo, L = Minato-ku, O = Sony Interactive Entertainment Inc., OU = Network Engineering & Operation Dept., CN = *.ww.np.dl.playstation.net

issuer=C = JP, O = Sony Computer Entertainment Inc., CN = SCEI DNAS Root 05

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 2785 bytes and written 445 bytes
Verification error: CA signature digest algorithm too weak
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:

(...)

    Start Time: 1615127012
    Timeout   : 7200 (sec)
    Verify return code: 68 (CA signature digest algorithm too weak)
    Extended master secret: no
---
 
closed
```
but I could not get it to work with this either.

**Comment by ioistired** [comment]
Huh. I would prefer to fix the certificate verification rather than ignore it. How did you install ps3udl?

**Comment by ukos-git** [comment]
I would also prefer to fix the certificate. It is not completely ignored, though. 

The second issue here is the location of the file.

 I installed from pip3. Can you reproduce the issues?

**Comment by ioistired** [comment]
I've pushed a fix for the second issue: 0f14fae53db2ebe067ba0aea2aeaee5a53dd3a4e. I still cannot reproduce the first one though. You seem to have the same certificate as I do, so I figure the certificate itself is fine. I wonder if there's some sort of OpenSSL configuration difference between our machines that's preventing verification.

**Comment by ukos-git** [comment]
Just a guess but the certificate could depend on the region/ip of the caller. Could you compare the openssl output on your side with the one that I posted?

**Comment by ioistired** [comment]
It's identical.

**Comment by autobalance** [comment]
Had the same issue, but noticed my system-wide openssl.cnf set `SECLEVEL=2`. Adding the line

`requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS = 'ALL:@SECLEVEL=1'`

fixes the issue, but I'm unsure if this is a good solution.

**Comment by ioistired** [comment]
That's a good lead, thanks. I don't know if there's a supported way to configure that though. I'll look into it.

**Comment by Desmo5574** [comment]
Was looking for an update downloader and came across this. Seems to be the only game update downloader for linux. Just to verify, I have the exact same error as ukos-git. 
I installed via pipx...

**Comment by telyn** [comment]
> Had the same issue, but noticed my system-wide openssl.cnf set `SECLEVEL=2`. Adding the line
> 
> `requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS = 'ALL:@SECLEVEL=1'`
> 
> fixes the issue, but I'm unsure if this is a good solution.

I used this technique - just dropped the line given by @autobalance in after the `import urllib3.connection` line in `ps3-update-dl.py`

Worked for me :)

**Comment by ukos-git** [comment]
I can confirm that the new package configuration `ps3-update-dl-0.0.1.post2` now finds the certificate at the right location. 

setting `SECLEVEL` as described above fixes the problem. I am not too deep into this topic so I cannot say if this is a safe setting to turn on. @ioistired this would close the issue and the ceritificate validation would *not* get skipped.

**Comment by ioistired** [comment]
Fixed in 0.0.2.
